File: misc/js/random_number_generator_in_a_range.js
// generates a random number in range [start, end]
function generateRandomNumber(start, end){
    return start + Math.floor(Math.random() * (end - start + 1))
}

function convertMapToArray(mp){
    const arr = []
    for(const [num, freq] of mp){
        arr.push({num, freq})
    }
    
    arr.sort((a, b) => a.num - b.num)
    
    return arr
}

let count = 518
const start = 1
const end = 50

const avgFreq = count / (end - start + 1)

// number -> frequency
const mp = new Map()

console.log(`Generating ${count} random numbers in range [${start}, ${end}]...\n`)

while(count-- > 0){
    const randNum = generateRandomNumber(start, end)
    
    if(mp.has(randNum)){
        mp.set(randNum, mp.get(randNum) + 1)
    }else{
        mp.set(randNum, 1)
    }
}

const arr = convertMapToArray(mp)

console.log("Number -> Frequency")
let minFreq = 2**31 - 1
let maxFreq = 0

for(const {num, freq} of arr){
    console.log(num + " -> " + freq)
    
    minFreq = Math.min(minFreq, freq)
    maxFreq = Math.max(maxFreq, freq)
}

console.log('\nMin freq = ' + minFreq)
console.log('Max freq = ' + maxFreq)
console.log('Avg freq expected = ' + avgFreq)

console.log('-----------------------------')

File: misc/cpp/random_number_generator_in_a_range.cpp
#include <bits/stdc++.h>
using namespace std;

// generates a random-number in range [start, end]
int generateRandomNumber(int start, int end){
    return start + rand() % (end - start + 1);
}

int main(){
    int count = 518;
    int start = 1;
    int end = 50;
    
    double avgFreq = (double)count / (end - start + 1);

    // number -> frequency
    map<int, int> mp;

    srand(time(0)); // seed for rand()
    
    cout<<"Generating "<<count<<" random numbers in range ["<<start<<", "<<end<<"]..."<<endl<<endl;
    
    while(count--)
        mp[generateRandomNumber(start, end)]++;
    
    cout<<"Number -> Frequency"<<endl;
    int minFreq = INT_MAX;
    int maxFreq = 0;
    
    for(auto [num, freq]: mp)
    {
        cout<<num<<" -> "<<freq<<endl;
        
        minFreq = min(minFreq, freq);
        maxFreq = max(maxFreq, freq);
    }
    
    cout<<endl;
    cout<<"Min Freq = "<<minFreq<<endl;
    cout<<"Max Freq = "<<maxFreq<<endl;
    cout<<"Avg Freq expected = "<<avgFreq<<endl;
    
    cout<<endl<<string(25, '-')<<endl;
    
    return 0;
}